### PR TITLE
PL011 driver

### PR DESCRIPTION
**Pull request by pj1031999**
Introduce _UART_ for _RPi3_.

Depends on fixes from #824, #825.

### CODE REVIEW

**Review by xThaid — APPROVED** [review]

File: sys/aarch64/gpio.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+/* The function select registers are used to define the operation of the
+ * general-purpose I/O pins. */
+#define GPFSEL(x) ((x) * sizeof(uint32_t))
+
+#define GPFSEL_PINS_PER_REGISTER 10
+#define GPFSEL_BITS_PER_PIN 3
+
+/* The output set registers are used to set a GPIO pin. */
+#define GPSET(x) (0x1C + (x) * sizeof(uint32_t))
```
Comment (cahirwpz) [inline sys/aarch64/gpio.c:14]: You have corresponding definitions here and in `bcm2835_gpioreg.h`. Please sort it out.

File: sys/aarch64/gpio.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+/* The asynchronous rising edge detect enable registers define the pins for
+ * which a asynchronous rising edge transition sets a bit in the event detect
+ * status registers GPEDS. */
+#define GPAREN(x) (0x7C + (x) * sizeof(uint32_t))
+
+/* The asynchronous falling edge detect enable registers define the pins for
+ * which a asynchronous falling edge transition sets a bit in the event detect
+ * status registers GPEDS. */
+#define GPAFEN(x) (0x88),
```
Comment (cahirwpz) [inline sys/aarch64/gpio.c:51]: This one is different that the others. Is that ok?

File: include/aarch64/bcm2835_gpioreg.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#define BCM2835_GPIO_GPPUD_PINS_PER_REGISTER 32
+
+#define BCM2838_GPIO_GPPUD_PULLOFF 0x0
+#define BCM2838_GPIO_GPPUD_PULLDOWN 0x2
+#define BCM2838_GPIO_GPPUD_PULLUP 0x1
+#define BCM2838_GPIO_GPPUPPDN(x) (0x0e4 + (x) * sizeof(uint32_t))
+#define BCM2838_GPIO_GPPUD_REGID(n) ((n) / 16)
+#define BCM2838_GPIO_GPPUD_MASK(n) (0x3 << ((n) % 16) * 2)
+
+/* brcm,function property */
```
Comment (cahirwpz) [inline include/aarch64/bcm2835_gpioreg.h:68]: I'd prefer to enclose it in an enum.

File: include/aarch64/gpio.h [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#ifndef _GPIO_H_
+#define _GPIO_H_
+
+struct resource_t;
+
+void delay(int64_t count);
```
Comment (cahirwpz) [inline include/aarch64/gpio.h:6]: Please don't expose it. You don't need to.

File: include/aarch64/gpio.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+void gpio_function_select(resource_t *r, unsigned pin, unsigned func);
```
Comment (cahirwpz) [inline include/aarch64/gpio.h:8]: Perhaps a short docstring would be helpful for readers. Could you link [Raspberry Pi Pinout](https://pinout.xyz) here for reference?

BTW. What is maximum value for `pin`. Should we check that with `assert`?

**Review by cahirwpz — COMMENTED** [review]
While `pl011` looks mostly ok, `gpio` stuff is kinda messy.

File: sys/aarch64/pl011.c [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  /* Mask all interrupts. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_ALLMASK);
+
+  /* Enable UART0, receive & transfer part of UART. */
+  bus_write_4(r, PL011COM_CR, PL01X_CR_UARTEN | PL011_CR_TXE | PL011_CR_RXE);
+
+  /* Enable receive interrupt. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_RX);
+
+  state->irq = bus_alloc_irq(dev, 0, BCM2835_INT_UART0, RF_ACTIVE);
```
Comment (cahirwpz) [inline sys/aarch64/pl011.c:226]: I rewrote interrupt allocation & handling using fresh pieces of driver infrastructure. Please check again if the driver works.

Comment (pj1031999) [reply sys/aarch64/pl011.c:226]: After some fixes.

File: include/aarch64/gpio.h (hunk 4 above)
Comment (pj1031999) [reply include/aarch64/gpio.h]: Done.

File: sys/aarch64/gpio.c (hunk 2 above)
Comment (pj1031999) [reply sys/aarch64/gpio.c]: Fixed.

File: sys/aarch64/gpio.c (hunk 1 above)
Comment (pj1031999) [reply sys/aarch64/gpio.c]: Done.

File: include/aarch64/bcm2835_gpioreg.h (hunk 3 above)
Comment (pj1031999) [reply include/aarch64/bcm2835_gpioreg.h:68]: Done.

File: include/aarch64/gpio.h (hunk 5 above)
Comment (pj1031999) [reply include/aarch64/gpio.h:9]: I added assert and links to documentation.

File: sys/aarch64/rootdev.c [hunk 7]
```diff
@@ -166,7 +166,9 @@ static void bcm2835_intr_handle(bus_space_handle_t irq_base, bus_size_t offset,
 
   while (pending) {
     int irq = ffs(pending) - 1;
-    intr_event_run_handlers(events[irq]);
+    /* XXX: some pending bits are shared between BASIC and GPU0/1. */
+    if (events[irq])
```
Comment (MichalBlk) [inline sys/aarch64/rootdev.c:170]: `intr_event_run_handlers` asserts that `ie != NULL`. If you get an interrupt and you don't have any associated
interrupt handler with it, you'll probably want to know it (and let it crush in the assert).

**Review by MichalBlk — APPROVED** [review]
LGTM

File: sys/aarch64/rootdev.c (hunk 7 above)
Comment (pj1031999) [reply sys/aarch64/rootdev.c:170]: Not in this case.
I really want to ignore duplicate interrupts here.

File: include/aarch64/gpio.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#ifndef _GPIO_H_
+#define _GPIO_H_
+
+struct resource_t;
+
+/* Select GPIO alt function. For more information look at
+ * https://cs140e.sergio.bz/docs/BCM2837-ARM-Peripherals.pdf and
+ * https://pinout.xyz */
+void gpio_function_select(resource_t *r, unsigned pin, unsigned func);
```
Comment (cahirwpz) [inline include/aarch64/gpio.h:9]: Should type of `func` be enum with `BCM2835_GPIO_*` ?

File: include/aarch64/gpio.h [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+void gpio_set_pull(resource_t *r, unsigned pin, unsigned pud);
```
Comment (cahirwpz) [inline include/aarch64/gpio.h:11]: Same here but enum with `BCM2838_GPIO_GPPUD_*`.

File: sys/aarch64/gpio.c [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                   BCM2835_GPIO_GPFSEL_BITS_PER_PIN;
+
+  unsigned val = bus_read_4(r, BCM2835_GPIO_GPFSEL(reg));
+  val &= ~(mask << shift);
+  val |= (func << shift);
+  bus_write_4(r, BCM2835_GPIO_GPFSEL(reg), val);
+}
+
+void gpio_set_pull(resource_t *r, unsigned pin, unsigned pud) {
+  assert(pin <= 40);
```
Comment (cahirwpz) [inline sys/aarch64/gpio.c:31]: Magic constant alert. Are pins counted from 0 or 1 ?

File: include/aarch64/gpio.h [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#ifndef _GPIO_H_
```
Comment (cahirwpz) [inline include/aarch64/gpio.h:1]: Should you add prefix `bcm2835_` for all the definitions below ?

File: include/aarch64/gpio.h (hunk 8 above)
Comment (pj1031999) [reply include/aarch64/gpio.h]: Done.

File: include/aarch64/gpio.h (hunk 9 above)
Comment (pj1031999) [reply include/aarch64/gpio.h]: Done.

File: include/aarch64/gpio.h (hunk 11 above)
Comment (pj1031999) [reply include/aarch64/gpio.h:1]: Done.

File: sys/aarch64/gpio.c (hunk 10 above)
Comment (pj1031999) [reply sys/aarch64/gpio.c]: It depends...
Now should be better.

**Review by cahirwpz — APPROVED** [review]
LGTM

File: sys/aarch64/pl011.c [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+  bus_write_4(r, PL011COM_LCRH, PL01X_LCR_FEN | PL01X_LCR_8BITS);
+
+  /* Mask all interrupts. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_ALLMASK);
+
+  /* Enable UART0, receive & transfer part of UART. */
+  bus_write_4(r, PL011COM_CR, PL01X_CR_UARTEN | PL011_CR_TXE | PL011_CR_RXE);
+
+  /* Enable receive interrupt. */
+  bus_write_4(r, PL011COM_IMSC, PL011_INT_RX);
```
Comment (ThomasStachovsky) [inline sys/aarch64/pl011.c:224]: In the [documentation](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2835/BCM2835-ARM-Peripherals.pdf) on the 177th page UARTRXINTR (PL011_INT_RX) is said to be associated with the transmit fifo and not receive fifo. They say UARTRTINTR (PL011_INT_RT) is the receive interrupt. I might not understand something. Can You comment on that?

**Review by ThomasStachovsky — COMMENTED** [review]
Reviewed with mohrcore. Left one comment regarding ```pl011.c```.

File: sys/aarch64/pl011.c (hunk 12 above)
Comment (mohrcore) [reply sys/aarch64/pl011.c:224]: As far as I get it, looking at the filter procedure for PL011, you want to enable both `UARTRXINTR` and `UARTRTINTR`.

File: sys/aarch64/pl011.c [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+  .v_ioctl = pl011_ioctl,
+  .v_getattr = pl011_getattr
+};
+/* clang-format on */
+
+static intr_filter_t pl011_intr(void *data /* device_t* */) {
+  pl011_state_t *state = ((device_t *)data)->state;
+  intr_filter_t res = IF_STRAY;
+
+  WITH_SPIN_LOCK (&state->lock) {
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/pl011.c:140]: No.
Look at `_spin_lock` – it calls `intr_disable`.

[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/aarch64/pl011.c:140]: Thank you for clarifying that. I havo no more remarks.

**Review by mohrcore — APPROVED** [review]
LGTM.

**Pull request merged**